### PR TITLE
DeepNorm, Noam

### PR DESCRIPTION
**Pull request by nehcream2od**
https://nehcream.notion.site/DeepNorm-3830dc4d2fa7455a81076706ce69a077?pvs=4

https://nehcream.notion.site/Noam-2d54370cf29f45d3a47679ce9e3ac16d?pvs=4

transformer : 트랜스포머
weight : 가중치
output : 출력/출력값
layer : 계층
normalization : 정규화
attention : 어텐션
sub-layer : 하위 계층
encoder : 인코더
decoder : 디코더
shape : 모양
flag : 플래그
embedding : 임베딩
module : 모듈
Auto-Regressive : 자기 회귀
Optimizer : 옵티마이저
learning rate : 학습률
bias : 편향

attention의 key, query, value : ?
causal mask : ?
scaling : ?
Readout : ?
logit : ?
warmup : ?

### CODE REVIEW

File: docs/ko/normalization/deep_norm/experiment.html [hunk 1]
```diff
@@ -0,0 +1,888 @@
...
+                    <pre><span class="lineno">55</span>        <span class="n">x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">emb</span><span class="p">(</span><span class="n">x</span><span class="p">)</span></pre>
+                </div>
+            </div>
+        </div>
+        <div class='section' id='section-9'>
+            <div class='docs'>
+                <div class='section-link'>
+                    <a href='#section-9'>#</a>
+                </div>
+                <p>트랜스포머 인코더입니다. </p>
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/experiment.html:234]: (제안) 토큰 임베딩에 트랜스포머를 적용합니다.

File: docs/ko/normalization/deep_norm/experiment.html [hunk 2]
```diff
@@ -0,0 +1,888 @@
...
+                    <pre><span class="lineno">57</span>        <span class="n">x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transformer</span><span class="p">(</span><span class="n">x</span><span class="p">)</span></pre>
+                </div>
+            </div>
+        </div>
+        <div class='section' id='section-10'>
+            <div class='docs'>
+                <div class='section-link'>
+                    <a href='#section-10'>#</a>
+                </div>
+                <p>logit 값을 가져옵니다. </p>
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/experiment.html:248]: (제안) logit 값을 구합니다.

File: docs/ko/normalization/deep_norm/experiment.html [hunk 3]
```diff
@@ -0,0 +1,888 @@
...
+                    <pre><span class="lineno">133</span>    <span class="n">experiment</span><span class="o">.</span><span class="n">create</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;deep_norm&quot;</span><span class="p">,</span> <span class="n">writers</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;screen&#39;</span><span class="p">,</span> <span class="s1">&#39;web_api&#39;</span><span class="p">})</span></pre>
+                </div>
+            </div>
+        </div>
+        <div class='section' id='section-27'>
+            <div class='docs'>
+                <div class='section-link'>
+                    <a href='#section-27'>#</a>
+                </div>
+                <p>설정을 생성합니다. </p>
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/experiment.html:613]: (제안) 설정 파일을 생성합니다.

File: docs/ko/normalization/deep_norm/experiment.html [hunk 4]
```diff
@@ -0,0 +1,888 @@
...
+                    <pre><span class="lineno">137</span>    <span class="n">experiment</span><span class="o">.</span><span class="n">configs</span><span class="p">(</span><span class="n">conf</span><span class="p">,</span> <span class="p">{</span></pre>
+                </div>
+            </div>
+        </div>
+        <div class='section' id='section-29'>
+            <div class='docs'>
+                <div class='section-link'>
+                    <a href='#section-29'>#</a>
+                </div>
+                <p>캐릭터 레벨 토크나이저를 사용합니다. </p>
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/experiment.html:641]: (제안) 글자 단위의 토크나이저를 사용합니다.

File: docs/ko/normalization/deep_norm/experiment.html [hunk 5]
```diff
@@ -0,0 +1,888 @@
...
+            </div>
+        </div>
+        <div class='section' id='section-33'>
+            <div class='docs'>
+                <div class='section-link'>
+                    <a href='#section-33'>#</a>
+                </div>
+                <p><span><span class="katex"><span aria-hidden="true" class="katex-html"><span class="base"><span
+                                        class="strut" style="height:0.64444em;vertical-align:0em;"></span><span
+                                        class="mord">256</span></span></span></span></span>의 문장 길이를 사용합니다. </p>
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/experiment.html:699]: (제안) 문장 길이는 256으로 설정합니다.

File: docs/ko/normalization/deep_norm/experiment.html [hunk 6]
```diff
@@ -0,0 +1,888 @@
...
+                    <pre><span class="lineno">148</span>        <span class="s1">&#39;seq_len&#39;</span><span class="p">:</span> <span class="mi">256</span><span class="p">,</span></pre>
+                </div>
+            </div>
+        </div>
+        <div class='section' id='section-34'>
+            <div class='docs'>
+                <div class='section-link'>
+                    <a href='#section-34'>#</a>
+                </div>
+                <p>32 에폭 훈련합니다. </p>
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/experiment.html:713]: (제안) 훈련 에폭은 32로 설정합니다.

File: docs/ko/normalization/deep_norm/experiment.html [hunk 7]
```diff
@@ -0,0 +1,888 @@
...
+                    <pre><span class="lineno">150</span>        <span class="s1">&#39;epochs&#39;</span><span class="p">:</span> <span class="mi">32</span><span class="p">,</span></pre>
+                </div>
+            </div>
+        </div>
+        <div class='section' id='section-35'>
+            <div class='docs'>
+                <div class='section-link'>
+                    <a href='#section-35'>#</a>
+                </div>
+                <p>배치 크기로 <span><span class="katex"><span aria-hidden="true" class="katex-html"><span class="base"><span
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/experiment.html:727]: (제안) 배치 크기는 16으로 설정합니다.

File: docs/ko/normalization/deep_norm/experiment.html [hunk 8]
```diff
@@ -0,0 +1,888 @@
...
+            </div>
+        </div>
+        <div class='section' id='section-36'>
+            <div class='docs'>
+                <div class='section-link'>
+                    <a href='#section-36'>#</a>
+                </div>
+                <p>한 에폭 당 <span><span class="katex"><span aria-hidden="true" class="katex-html"><span class="base"><span
+                                        class="strut" style="height:0.64444em;vertical-align:0em;"></span><span
+                                        class="mord">10</span></span></span></span></span>회 훈련과 검증을 전환합니다. </p>
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/experiment.html:745]: (제안) 각 에폭에서 훈련과 검증 전환을 10회 수행합니다.

File: docs/ko/normalization/deep_norm/index.html [hunk 9]
```diff
@@ -0,0 +1,1685 @@
...
+                        href="https://colab.research.google.com/github/labmlai/annotated_deep_learning_paper_implementations/blob/master/labml_nn/normalization/deep_norm/experiment.ipynb"><img
+                            alt="Open In Colab" src="https://colab.research.google.com/assets/colab-badge.svg"></a></p>
+                <p>이 페이지는 <a href="https://papers.labml.ai/paper/2203.00555">DeepNet: Scaling Transformers to 1,000
+                        Layers</a> 논문에서 DeepNorm을 <a href="https://pytorch.org">PyTorch</a>로 구현한 내용에 대한 것입니다.</p>
+                <p>이 논문은 계층 정규화를 대체하는 새로운 정규화 함수와 가중치 초기화 방법을 통해 극도로 깊은 트랜스포머를 안정화하는 방법을
+                    제안합니다. DeepNorm은 Post-LayerNorm의 성능과 Pre-LayerNorm의 안정성을 결합하였습니다. DeepNorm이 적용된 트랜스포머는
+                    학습률의 warm-up[워밍업]없이도 안정적입니다.</p>
+                <p>이 논문은 먼저 동일한 입력에 대한 계층 출력값의 변화가 안정적인 훈련에서는 점진적으로 변화하고, 불안정한 경우 훈련 초기에 급격하게
+                    변화한다는 것을 보여줍니다. 이는 가중치를 작은 값으로 초기화하거나 훈련이 안정된 상태에서 학습률을 warm-up 할 때 발생합니다. 연구진은 계층 출력값의 변화를 작게 유지한다는
+                    아이디어에서 새로운 정규화 및 가중치 초기화의 메커니즘을 도출해 냈습니다.</p>
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/index.html:91]: (제안) 레이어 출력값의 변화를 작게 유지함으로써 새로운 정규화 및 가중치 초기화 매커니즘을 유도하고자 하였습니다.

File: docs/ko/normalization/deep_norm/index.html [hunk 10]
```diff
@@ -0,0 +1,1685 @@
...
+                            class="katex"><span aria-hidden="true" class="katex-html"><span class="base"><span
+                                        class="strut"
+                                        style="height:0.8888799999999999em;vertical-align:-0.19444em;"></span><span
+                                        class="mord coloredeq eqg" style=""><span class="mord mathnormal"
+                                            style="margin-right:0.05278em">β</span></span></span></span></span></span>로
+                    조절합니다.
+
+                </p>
+                <p>DeepNorm은 <a href="../../transformers/feed_forward.html">순방향 신경망</a>과
+                    어텐션의 value 및 출력 계층의 가중치 스케일링을 제안합니다. 이러한 변환의 가중치들은 <span><span class="katex"><span
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/index.html:103]: (제안) 가중치를 어떻게 스케일링 할지는 beta에 의해 결정됩니다.

File: docs/ko/normalization/deep_norm/index.html [hunk 11]
```diff
@@ -0,0 +1,1685 @@
...
+            </div>
+        </div>
+        <div class='section' id='section-8'>
+            <div class='docs doc-strings'>
+                <div class='section-link'>
+                    <a href='#section-8'>#</a>
+                </div>
+                <ul>
+                    <li><code class="highlight"><span></span><span class="n">d_model</span></code>
+                        은 토근의 임베딩 크기입니다. </li>
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/index.html:1412]: 토큰

File: docs/ko/normalization/deep_norm/index.html [hunk 12]
```diff
@@ -0,0 +1,1685 @@
...
+                        은 셀프 어텐션 모듈입니다. </li>
+                    <li><code class="highlight"><span></span><span class="n">feed_forward</span></code>
+                        는 순방향 신경망 모듈입니다. </li>
+                    <li><code class="highlight"><span></span><span class="n">deep_norm_alpha</span></code>
+                        는 DeepNorm의 <span><span class="katex"><span aria-hidden="true" class="katex-html"><span
+                                        class="base"><span class="strut"
+                                            style="height:0.43056em;vertical-align:0em;"></span><span
+                                            class="mord coloredeq eqf" style=""><span class="mord mathnormal"
+                                                style="margin-right:0.0037em">α</span></span></span></span></span></span>
+                        계수입니다. </li>
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/index.html:1423]: (제안) 계수 \alpha 입니다.

File: docs/ko/normalization/deep_norm/index.html [hunk 13]
```diff
@@ -0,0 +1,1685 @@
...
+                    <li><code class="highlight"><span></span><span class="n">deep_norm_beta</span></code>
+                        는 DeepNorm의 가중치 초기화 스케일링을 위한 <span><span class="katex"><span aria-hidden="true"
+                                    class="katex-html"><span class="base"><span class="strut"
+                                            style="height:0.8888799999999999em;vertical-align:-0.19444em;"></span><span
+                                            class="mord coloredeq eqg" style=""><span class="mord mathnormal"
+                                                style="margin-right:0.05278em">β</span></span></span></span></span></span>
+                        상수입니다.</li>
```
Comment (Bandi120424) [inline docs/ko/normalization/deep_norm/index.html:1430]: (제안) 상수 \beta 입니다.

File: docs/ko/optimizers/noam.html [hunk 14]
```diff
@@ -0,0 +1,870 @@
...
+                                                                    class="pstrut" style="height:2.7em;"></span><span
+                                                                    class="sizing reset-size6 size3 mtight"><span
+                                                                        class="mord mtight">2</span></span></span></span><span
+                                                            class="vlist-s">​</span></span><span class="vlist-r"><span
+                                                            class="vlist"
+                                                            style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span></span>)
+                        의 튜플입니다.
+                    </li>
+                    <li><code class="highlight"><span></span><span class="n">eps</span></code>
+                        은 <code class="highlight"><span></span><span class="n">optimized_update</span></code>에 기반한
```
Comment (Bandi120424) [inline docs/ko/optimizers/noam.html:158]: (제안) optimized_update 값에 따라 \hat\epsilon 또는 \epsilon의 값을 갖습니다.

File: docs/ko/optimizers/noam.html [hunk 15]
```diff
@@ -0,0 +1,870 @@
...
+                    <li>&#x27;optimized_update&#x27;는 <span><span class="katex"><span aria-hidden="true"
+                                    class="katex-html"><span class="base"><span class="strut"
+                                            style="height:0.43056em;vertical-align:0em;"></span><span
+                                            class="mord coloredeq eqd" style=""><span class="mord mathnormal"
+                                                style="">ϵ</span></span></span></span></span></span>을 더한 후 두 번째 모멘트의 편향
+                        보정을 최적화할지 여부를 정하는 플래그입니다.</li>
+                    <li><code class="highlight"><span></span><span class="n">amsgrad</span></code>
+                        는 AMSGrad를 사용할지, 일반 Adam을 사용할지를 지정하는 플래그입니다. </li>
+                    <li><code class="highlight"><span></span><span class="n">warmup</span></code>
+                        은 warmup 스텝입니다. </li>
```
Comment (Bandi120424) [inline docs/ko/optimizers/noam.html:193]: (제안) warmup 스텝의 개수

**Comment by nehcream2od** [comment]
> 번역하신 것 보고 저도 질문이 생겨 코멘트 남깁니다 😃
> 
> 1. 저는 **shape**을 **형상**으로 번역하였는데 **모양**으로 통일할까요?
> 2. (굉장히 사소한 문제기는 하지만) head는 그대로 영어로 남기는지 아니면 헤드라고 쓰는지요?
> 3. configuration이나 config는 어떻게 번역하면 좋을까요? 그대로 두는 것이 나을까요?
> 
> * 저도 layer를 계층으로 번역하였는데 의견을 따라 "레이어"로 통일하겠습니다.

1. 앗 저는 형상이란 말이 좀 더 와닿는 것 같은데 이따 다른분들이랑도 얘기해보면 좋을 것 같습니다

2. 번역 작업이니까 이왕이면 헤드라고 바꾸는게 낫지 않을까 싶습니다!

3. 설정이란 말이 적합하지 않을까요?

세세한 피드백 감사합니다 말씀하신 사항들 이따 수정해놓도록 하겠습니다 ㅎㅎ

**Comment by nehcream2od** [comment]
> attention의 K, Q, V는 그냥 각각 키, 쿼리, 밸류 scaling: 스케일링 logit: 로짓 warmup: 웜업
> 
> 그대로 들고 가셔도 무방할 것 같다는 생각이 들어요 :)
> 
> causal mas: 인과적 마스크? Readout은 문맥에 따라 적절한 선택이 필요하지 않을까요?
> 
> +) 저는 "layer" 쪽 번역은 "계층" 보다는 "레이어"가 낫다고 생각합니다. 첫 번째 이유는 ML/DL 분야에서 "hierarchical" 용어가 "계층~" 의미로 더 널리 쓰이고 있어서 혼동을 줄 수 있고요. 두 번째 이유는 이 분야를 보는 사람이라면 "layer" 말 자체를 그냥 "레이어"로 더 잘 받아들일 것 같아서요. (어텐션의 키, 쿼리, 밸류를 그대로 말하는 것과 같은 이유입니다.)

말씀하신 부분을 생각해보니 레이어라고 두는게 적합할 것 같네요! 피드백 감사합니당 ㅎㅎ

**Comment by movie5** [comment]
감사합니다 👍🏼
